dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2099 @@
- */
-HRESULT WINAPI CAutoComplete::GetOptions(DWORD *pdwFlag)
+// WM_KEYDOWN
+// This message is posted to the window with the keyboard focus when a non-system key is pressed.
+LRESULT CACEditCtrl::OnKeyDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
 {
-    HRESULT hr = S_OK;
+    TRACE("CACEditCtrl::OnKeyDown(%p, %p)\n", this, wParam);
+    ATLASSERT(m_pDropDown);
+    if (!m_pDropDown->OnEditKeyDown(wParam, lParam))
+        return DefWindowProcW(uMsg, wParam, lParam); // do default
+    return 1;
+}
 
-    TRACE("(%p) -> (%p)\n", this, pdwFlag);
+// WM_KILLFOCUS @implemented
+// This message is sent to a window immediately before it loses the keyboard focus.
+LRESULT CACEditCtrl::OnKillFocus(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnKillFocus(%p)\n", this);
+    ATLASSERT(m_pDropDown);
 
-    *pdwFlag = m_options;
+    // hide the list if lost focus
+    HWND hwndGotFocus = (HWND)wParam;
+    if (hwndGotFocus != m_hWnd && hwndGotFocus != m_pDropDown->m_hWnd)
+    {
+        m_pDropDown->HideDropDown();
+    }
 
-    return hr;
+    return DefWindowProcW(uMsg, wParam, lParam); // do default
 }
 
-/**************************************************************************
- *  IAutoComplete_fnSetOptions
- */
-HRESULT WINAPI CAutoComplete::SetOptions(DWORD dwFlag)
+// WM_PASTE @implemented
+// An application sends this message to an edit control to paste at the current caret position.
+LRESULT CACEditCtrl::OnPaste(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
 {
-    HRESULT hr = S_OK;
-
-    TRACE("(%p) -> (0x%x)\n", this, dwFlag);
-
-    m_options = (AUTOCOMPLETEOPTIONS)dwFlag;
+    TRACE("CACEditCtrl::OnPaste(%p)\n", this);
+    ATLASSERT(m_pDropDown);
 
-    if ((m_options & ACO_AUTOSUGGEST) && m_hwndEdit && !m_hwndListBox)
-        CreateListbox();
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // do default
+    m_pDropDown->OnEditUpdate(TRUE);
+    return ret;
+}
 
-    return hr;
+// WM_SETFOCUS
+// This message is sent to a window after it has gained the keyboard focus.
+LRESULT CACEditCtrl::OnSetFocus(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnSetFocus(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    return DefWindowProcW(uMsg, wParam, lParam); // do default
 }
 
-/* Edit_BackWord --- Delete previous word in text box */
-static void Edit_BackWord(HWND hwndEdit)
+// WM_SETTEXT
+// An application sends this message to set the text of a window.
+LRESULT CACEditCtrl::OnSetText(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
 {
-    INT iStart, iEnd;
-    iStart = iEnd = 0;
-    SendMessageW(hwndEdit, EM_GETSEL, (WPARAM)&iStart, (LPARAM)&iEnd);
+    TRACE("CACEditCtrl::OnSetText(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    if (!m_pDropDown->m_bInSetText)
+        m_pDropDown->HideDropDown(); // it's mechanical WM_SETTEXT
+    return DefWindowProcW(uMsg, wParam, lParam); // do default
```

Comment:
I think that if you set bHandled to false the caller will automatically call DefWindowProcW

Reply:
OK.